File: src/test/java/com/severett/kotlinintro/controller/OfferedServiceControllerIT.java
package com.severett.kotlinintro.controller;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.severett.kotlinintro.model.OfferedService;
import jakarta.transaction.Transactional;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.mock.web.MockHttpServletResponse;
import org.springframework.test.web.servlet.MockMvc;

import java.math.BigDecimal;
import java.util.HashMap;

import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.put;

@SpringBootTest
@AutoConfigureMockMvc
public class OfferedServiceControllerIT {
    private static final String CHECKUP_NAME = "Checkup";
    private static final BigDecimal CHECKUP_PRICE = new BigDecimal("25.50");

    private final ObjectMapper objectMapper = new ObjectMapper();

    private final MockMvc mvc;

    public OfferedServiceControllerIT(@Autowired MockMvc mvc) {
        this.mvc = mvc;
    }

    @Test
    public void getOfferedServiceTest() throws Exception {
        var response = mvc.perform(get("/offered_services/1")).andReturn().getResponse();
        assertEquals(HttpStatus.OK.value(), response.getStatus());
        var body = objectMapper.readValue(response.getContentAsString(), OfferedService.class);
        checkDish(body, CHECKUP_NAME, CHECKUP_PRICE, new BigDecimal("0.00"));
    }

    @Test
    public void getOfferedServiceNotFoundTest() throws Exception {
        var response = mvc.perform(get("/offered_services/2000")).andReturn().getResponse();
        assertAll(
                () -> assertEquals(HttpStatus.NOT_FOUND.value(), response.getStatus()),
                () -> assertEquals(0, response.getContentLength())
        );
    }

    @Test
    @Transactional
    public void createOfferedServiceTest() throws Exception {
        var offeredServiceName = "Microchip Injection";
        var price = new BigDecimal("15.00");
        var bodyMap = new HashMap<String, Object>();
        bodyMap.put("name", offeredServiceName);
        bodyMap.put("price", price);

        var response = mvc.perform(
                post("/offered_services")
                        .content(objectMapper.writeValueAsBytes(bodyMap))
                        .contentType(MediaType.APPLICATION_JSON)
        ).andReturn().getResponse();

        assertEquals(HttpStatus.CREATED.value(), response.getStatus());
        var body = objectMapper.readValue(response.getContentAsString(), OfferedService.class);
        checkDish(body, offeredServiceName, price, BigDecimal.ZERO);
    }

    @Test
    @Transactional
    public void setDiscountTest() throws Exception {
        var discount = new BigDecimal("5.50");

        var response = mvc.perform(
                put("/offered_services/1/setDiscount?discount=" + discount)
        ).andReturn().getResponse();

        assertEquals(HttpStatus.OK.value(), response.getStatus());
        var body = objectMapper.readValue(response.getContentAsString(), OfferedService.class);
        checkDish(body, CHECKUP_NAME, CHECKUP_PRICE, discount);
    }

    @Test
    @Transactional
    public void setDiscountNotFoundTest() throws Exception {
        MockHttpServletResponse response = mvc.perform(
                put("/offered_services/2000/setDiscount?discount=5.50")
        ).andReturn().getResponse();
        assertAll(
                () -> assertEquals(HttpStatus.NOT_FOUND.value(), response.getStatus()),
                () -> assertEquals(0, response.getContentLength())
        );
    }

    private void checkDish(OfferedService offeredService, String name, BigDecimal price, BigDecimal discount) {
        assertAll(
                () -> assertEquals(name, offeredService.getName()),
                () -> assertEquals(price, offeredService.getPrice()),
                () -> assertEquals(discount, offeredService.getDiscount())
        );
    }
}

File: src/test/kotlin/com/severett/kotlinintro/controller/OfferedServiceControllerIT.kt
package com.severett.kotlinintro.controller

import com.fasterxml.jackson.databind.ObjectMapper
import com.severett.kotlinintro.model.OfferedService
import jakarta.transaction.Transactional
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertAll
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.put
import java.math.BigDecimal

private const val CHECKUP_NAME = "Checkup"
private val CHECKUP_PRICE = BigDecimal("25.50")

@SpringBootTest
@AutoConfigureMockMvc
class OfferedServiceControllerIT(@Autowired private val mvc: MockMvc) {
    private val objectMapper = ObjectMapper()

    @Test
    fun getOfferedServiceTest() {
        val response = mvc.perform(get("/offered_services/1")).andReturn().response
        assertEquals(HttpStatus.OK.value(), response.status)
        val body = objectMapper.readValue(response.contentAsString, OfferedService::class.java)
        checkOfferedService(body, CHECKUP_NAME, CHECKUP_PRICE, BigDecimal("0.00"))
    }

    @Test
    fun getOfferedServiceNotFoundTest() {
        val response = mvc.perform(get("/offered_services/2000")).andReturn().response
        assertAll(
            { assertEquals(HttpStatus.NOT_FOUND.value(), response.status) },
            { assertEquals(0, response.contentLength) },
        )
    }

    @Test
    @Transactional
    fun createOfferedServiceTest() {
        val offeredServiceName = "Microchip Injection"
        val price = BigDecimal("15.00")
        val bodyMap = buildMap {
            put("name", offeredServiceName)
            put("price", price)
        }
        val response = mvc.perform(
            post("/offered_services")
                .content(objectMapper.writeValueAsBytes(bodyMap))
                .contentType(MediaType.APPLICATION_JSON)
        ).andReturn().response
        assertEquals(HttpStatus.CREATED.value(), response.status)
        val body = objectMapper.readValue(response.contentAsString, OfferedService::class.java)
        checkOfferedService(body, offeredServiceName, price, BigDecimal.ZERO)
    }

    @Test
    @Transactional
    fun setDiscountTest() {
        val discount = BigDecimal("5.50")
        val response = mvc.perform(put("/offered_services/1/setDiscount?discount=$discount"))
            .andReturn().response
        assertEquals(HttpStatus.OK.value(), response.status)
        val body = objectMapper.readValue(response.contentAsString, OfferedService::class.java)
        checkOfferedService(body, CHECKUP_NAME, CHECKUP_PRICE, discount)
    }

    @Test
    @Transactional
    fun setDiscountNotFoundTest() {
        val response = mvc.perform(put("/offered_services/2000/setDiscount?discount=5.50"))
            .andReturn().response
        assertAll(
            { assertEquals(HttpStatus.NOT_FOUND.value(), response.status) },
            { assertEquals(0, response.contentLength) },
        )
    }

    private fun checkOfferedService(
        offeredService: OfferedService,
        name: String,
        price: BigDecimal,
        discount: BigDecimal
    ) {
        assertAll(
            { assertEquals(name, offeredService.name) },
            { assertEquals(price, offeredService.price) },
            { assertEquals(discount, offeredService.discount) },
        )
    }
}
